src/snowflake/snowpark/session.py

```diff
@@ -4513,3 +4576,53 @@ def _explain_query(self, query: str) -> Optional[str]:
             return None
 
     createDataFrame = create_dataframe
+
+    def _execute_sproc_internal(
+        self,
+        query: str,
+        is_return_table: bool,
+        block: bool,
+        statement_params: Optional[Dict[str, Any]] = None,
+        ast_stmt=None,
+        log_on_exception: bool = False,
+    ) -> Union[Any, AsyncJob]:
+        """Unified internal executor for sync/async, table/scalar SPROCs."""
+        if is_return_table:
+            if block:
+                qid = self._conn.execute_and_get_sfqid(
+                    query, statement_params=statement_params
+                )
+                df = self.sql(result_scan_statement(qid), _ast_stmt=ast_stmt)
+                set_api_call_source(df, "Session.call")
+                return df
+            else:
+                results_cursor = self._conn.execute_async_and_notify_query_listener(
+                    query, _statement_params=statement_params
+                )
+                return AsyncJob(
+                    results_cursor["queryId"],
+                    query,
+                    self,
+                    _AsyncResultType.ROW,
+                    log_on_exception=log_on_exception,
+                )
+        else:
+            if block:
+                # TODO SNOW-1672561: This here needs to emit an eval as well.
+                df = self.sql(query, _ast_stmt=ast_stmt)
+                set_api_call_source(df, "Session.call")
+                # Note the collect is implicit within the stored procedure call, so should not emit_ast here.
+                return df.collect(statement_params=statement_params, _emit_ast=False)[
+                    0
+                ][0]
+            else:
+                results_cursor = self._conn.execute_async_and_notify_query_listener(
+                    query, _statement_params=statement_params
+                )
+                return AsyncJob(
+                    results_cursor["queryId"],
+                    query,
+                    self,
+                    _AsyncResultType.COUNT,
+                    log_on_exception=log_on_exception,
+                )
```

Comment:
nit: I feel there's some redundancy in this part of code, for the async part, as far as I can see the only difference is _AsyncResultType.COUNT vs _AsyncResultType.ROW

e.g.,

the returning can just be

```python
                results_cursor = self._conn.execute_async_and_notify_query_listener(
                    query, _statement_params=statement_params
                )
                return AsyncJob(
                    results_cursor["queryId"],
                    query,
                    self,
                    _AsyncResultType. ROW if is_return_table else _AsyncResultType.COUNT,
                    log_on_exception=log_on_exception,
                )
```